src/autoschedulers/anderson2021/DefaultCostModel.cpp

```diff
@@ -0,0 +1,401 @@
+// This file is a wrapper around the cost model that loads and saves
+// weights, and maintains state of various kinds. For the actual cost
+// model, see cost_model_generator.cpp
+
+#include <algorithm>
+#include <cmath>
+#include <ctime>
+#include <map>
+#include <random>
+#include <sstream>
+#include <string>
+
+#include "ASLog.h"
+#include "DefaultCostModel.h"
+#include "HalideBuffer.h"
+#include "NetworkSize.h"
+#include "anderson2021_cost_model.h"
+#include "anderson2021_train_cost_model.h"
+
+// This is an embedded version of `baseline.weights`.
+// The embedding is done using binary2cpp.
+
+extern "C" unsigned char baseline_weights[];
+extern "C" int baseline_weights_length;
+
+namespace Halide {
+namespace {
+
+using Halide::Internal::aslog;
+using Halide::Internal::PipelineFeatures;
+using Halide::Internal::ScheduleFeatures;
+using Halide::Internal::Weights;
+using Halide::Runtime::Buffer;
+
+bool ends_with(const std::string &str, const std::string &suffix) {
+    if (str.size() < suffix.size()) return false;
+    size_t off = str.size() - suffix.size();
+    for (size_t i = 0; i < suffix.size(); i++) {
+        if (str[off + i] != suffix[i]) return false;
+    }
+    return true;
+}
+
+}  // namespace
+
+void DefaultCostModel::set_pipeline_features(const Internal::Autoscheduler::FunctionDAG &dag,
+                                             const Internal::Autoscheduler::Anderson2021Params &params) {
+
+    const int pipeline_feat_size = head1_w * head1_h;
+    // We ignore the first seven pipeline features in the cost
+    // model. It's just a mask of which types are in use.
+    static_assert(sizeof(PipelineFeatures) - 7 * sizeof(int) ==
+                      sizeof(int) * pipeline_feat_size,
+                  "Incorrect size for pipeline features");
+    int num_stages = 0;
+    for (const auto &n : dag.nodes) {
+        if (!n.is_input) num_stages += (int)n.stages.size();
```

Comment:
style nit, here and elsewhere: generally, Halide always encloses if clauses in braces, even for a single line:

```
if (...) {
   ...
}
```